Tachyon oauth

geekingfrog opened the pull request:
This implements the two oauth flows required for running tachyon.

There's also at the time of writing, a demo server running at [https://tachyon.geekingfrog.com:4567/login](https://tachyon.geekingfrog.com:4567)

There's currently only one OAuth application registered, with the hardcoded scope `tachyon.lobby`, one redirect uri: `http://127.0.0.1/oauth2callback` and the client id is `generic_lobby`.

# The lobby flow
This is using the `authorization_code` flow.

## getting an authorization code
First, the client need to [generate a PKCE](https://www.rfc-editor.org/rfc/rfc7636#page-8) verifier and challenge.
For example:
```
challenge: "BGLMtLONQ_f6-Z6ikTk8ofWo-cWM3UUeT93LIEG33-M"
verifier: "2ENOENOGA0USUNPROMSUD9U64P604R2LVOVDG5SEL7EIGA5SL3TC2BQN0MJVVG8S"
```

Then, the client can request an authorization code with a `GET` request to the endpoint:

https://tachyon.geekingfrog.com:4567/oauth/authorize?response_type=code&code_challenge=BGLMtLONQ_f6-Z6ikTk8ofWo-cWM3UUeT93LIEG33-M&code_challenge_method=S256&client_id=generic_lobby&redirect_uri=http%3A%2F%2F127.0.0.1%2Foauth2callback

You get redirected to a login screen.
The demo server has one user with email/password: `tachyon@foo.bar` and password: `tachyonmelon`

You then should see a screen to grant access:

![2024-06-16-856x269-scrot](https://github.com/beyond-all-reason/teiserver/assets/1531763/7b045ea9-0115-48a2-8aad-790ca79d354f)

Clicking on "Let's go!" redirects to: `127.0.0.1/oauth2callback?code=<code>`

## exchanging the authorization code for an access token
You can then issue a `POST` request to

```
curl -iv https://tachyon.geekingfrog.com:4567/oauth/token \
--data-urlencode "grant_type=authorization_code" \
--data-urlencode "client_id=generic_lobby" \
--data-urlencode "code=60PKPAAVGIL8L6MCSVQK90V1GLORV22PQI6H13PIGQL21E0HFAF0====" \
--data-urlencode "code_verifier=2ENOENOGA0USUNPROMSUD9U64P604R2LVOVDG5SEL7EIGA5SL3TC2BQN0MJVVG8S" \
--data-urlencode "redirect_uri=http://localhost/oauth2callback"
```

and you get back
```json
{
  "access_token":"C45B3IG4CAHUDG1TAN03USAA55KDS2NA30VQIFLTN9FNMICPA2DG",
  "expires_in":1800,
  "refresh_token":"15IA4J2O0NCH08D6URPK6O3S38G3B7RFBFOPJVV20MNJNHJ11JIG",
  "token_type":"Bearer"
}
```

# the autohost flow

I configured one client_id/client_secret pair that can be used to retrieve an auth token:

```bash
curl -ivv "https://tachyon.geekingfrog.com:4567/oauth/token" \
--data-urlencode "grant_type=client_credentials" \
--data-urlencode "client_id=207966ee-6997-433f-a1fa-28da8f3b754e" \
--data-urlencode "client_secret=NENEGFN5MNNA6TH34P53JE88STHA5RCQJ06RKTG5HN92UMAOQ72G"
```

and this gives the same type of response.

# Testing
There are automated tests under 2 location, one for the context and one for the controllers:
```
mix test test/teiserver/o_auth/
mix test test/teiserver_web/controllers/o_auth
```

# Out of (OAuth) Scope for this PR

* Currently the only thing you can do with an auth token is refresh it. I'll implement the basic websocket connection later.
* Any CRUD operations or interface to manage applications and client credentials. For now, manually inserting things in the DB will have to do.

p2004a commented:
:tada: I will review it for sure (don't have perms to assign myself as reviewer)

p2004a commented on lib/teiserver_web/views/api/o_auth/code_view.ex:
```diff
@@ -0,0 +1,35 @@
+      issuer: base,
+      authorization_endpoint: base <> ~p"/oauth/authorize",
+      token_endpoint: base <> ~p"/oauth/token",
+      token_endpoint_auth_methods_supported: ["none", "client_secret_post"],
```
https://www.rfc-editor.org/rfc/rfc6749.html#section-2.3.1

> The authorization server MUST support the HTTP Basic authentication scheme for authenticating clients that were issued a client password.

I interpret that as `client_secret_basic` effectively being required on this list.

geekingfrog replied:
I've added support for basic auth in 19bc2c23d955ced5a6fab5f71d7fc0bc05e77bcd

p2004a replied:
This seems to not be resolved, I've tested with curl and I get `missing client_id` when passing client_id via Basic auth:

```
curl --user generic_lobby: -d grant_type=authorization_code -s -X POST http://localhost:4000/oauth/token
```

p2004a commented on lib/teiserver_web/controllers/o_auth/authorize_controller.ex:
```diff
@@ -0,0 +1,104 @@
+
+        case OAuth.create_code(conn.assigns.current_user, code_params) do
+          {:error, _} ->
+            error_redirect(conn, redir_url, "server_error", "something went wrong")
```
`state` is also required in error response if it was present in the request.

geekingfrog replied:
I've changed the [error_redirect function](https://github.com/geekingfrog/teiserver/blob/3514c247d065f9b028c50923e390ecc919d9513f/lib/teiserver_web/controllers/o_auth/authorize_controller.ex#L110-L130) to make it so.

p2004a commented on lib/teiserver_web/controllers/o_auth/code_controller.ex:
```diff
@@ -0,0 +1,96 @@
+        exchange_token(conn, params)
+
+      missing_key ->
+        conn |> put_status(400) |> render(:error, error_description: "missing #{missing_key}")
```
Doesn't like like error is generated according to https://www.rfc-editor.org/rfc/rfc6749.html#section-4.1.4 here and below ?

geekingfrog replied:
I've amended it to return `{"error": "invalid_request"}` [at least](https://github.com/geekingfrog/teiserver/blob/3514c247d065f9b028c50923e390ecc919d9513f/test/teiserver_web/controllers/o_auth/code_controller_test.exs#L104).

p2004a commented on lib/teiserver/o_auth.ex:
```diff
@@ -0,0 +1,386 @@
+        value: Base.hex_encode32(:crypto.strong_rand_bytes(32), padding: false),
+        application_id: application.id,
+        scopes: application.scopes,
+        expires_at: Timex.add(now, Timex.Duration.from_days(60)),
```
I think this should be *much* longer then 60, IMHO it should be almost forever: https://discord.com/channels/549281623154229250/724293659809284137/1162856143719243888

geekingfrog replied:
Amended and bumped it to 100 years.

p2004a reviewed (commented):
How does it work with the authorization screen when user is not logged in? Will it first ask to login, then forward directly to auth screen?

geekingfrog commented on lib/teiserver_web/templates/o_auth/authorize/authorize.html.heex:
```diff
@@ -0,0 +1,37 @@
+          <button type="submit" class="btn btn-primary btn-block">Let's go!</button>
+        </form>
+        <br />
+        <a href={~p"/"}>
```
good spot, I forgot to return to that. I've added [the correct redirect uri](https://github.com/geekingfrog/teiserver/blob/3514c247d065f9b028c50923e390ecc919d9513f/lib/teiserver_web/controllers/o_auth/authorize_controller.ex#L20-L34) as the target there.

geekingfrog commented on lib/teiserver/o_auth/schemas/token.ex:
```diff
@@ -0,0 +1,40 @@
+    belongs_to :application, OAuth.Application, primary_key: true
+    field :scopes, {:array, :string}
+    field :expires_at, :utc_datetime
+    field :type, Ecto.Enum, values: [:bearer, :refresh]
```
done in the amended commit.

geekingfrog commented:
> How does it work with the authorization screen when user is not logged in? Will it first ask to login, then forward directly to auth screen?

Yes exactly. It'll preserve any potential querystring (so state stuff is there).

geekingfrog commented:
Thanks for the review p2004a, I believe I've addressed all your comments. Ping me on discord if you want me to restart the test server on my (tiny) vps, I've turned it off to save a bit of resources.

p2004a commented:
I've done a bunch of testing today, the biggest problem is lack of support for mandatory HTTP Basic authentication scheme for all but client credentials grant, that's a blocker.

A bunch of other things I've noticed that I don't think are hard blockers:
1. `state` is always passed to redirect uri, even when not set in the initial request at all.

   That's on the edge of spec correctness AFAIU.
3. Refresh token shouldn't be generated for client credentials grant: https://datatracker.ietf.org/doc/html/rfc6749#section-4.4.3

   But shouldn't != must not.
4. invalid scopes not verified properly, overall I'm not sure how well the scopes are handled. E.g. sending `curl -d grant_type=refresh_token -d refresh_token=N0G2J8BN1TTFA5HGGLGPMG1PR2V9EJ2LFT8GGO8G8MAR5K1BG6JG -d scope=asdasd -s -X POST http://localhost:4000/oauth/token` doesn't error out.

   That is incorrect per spec, but unlikely any clients at the moment will run into it.
6. The error codes are *very* vague, `invalid_request` is returned very often when a different error code like `invalid_grant` should be user or even `invalid_client` when bad credentials are passed in client credentials grant. `error_description` is often not filled in.

   That just makes it harder to debug interaction with server for clients.

---

One more question: how are old codes/tokens etc cleaned up from db?

p2004a commented:
One more question more on design side: I'm not sure I understand the relationship between autohost and application, why they are tied, can you elaborate on that design choice?

geekingfrog commented:
> One more question more on design side: I'm not sure I understand the relationship between autohost and application, why they are tied, can you elaborate on that design choice?

I don't think an autohost is tied to an application. [The schema](https://github.com/beyond-all-reason/teiserver/pull/335/files#diff-70aba1e028c3f1f94c457ff3468f3ce73966797913a16670fbdd3ae0d4bdf64bR11-R15) only specify an id and a name (and some timestamps).
The tokens and client id/secret are the thing tied to an application, for the authorized scopes and whatnot.

p2004a commented:
> I don't think an autohost is tied to an application. [The schema](https://github.com/beyond-all-reason/teiserver/pull/335/files#diff-70aba1e028c3f1f94c457ff3468f3ce73966797913a16670fbdd3ae0d4bdf64bR11-R15) only specify an id and a name (and some timestamps).
The tokens and client id/secret are the thing tied to an application, for the authorized scopes and whatnot.

Ah, `application <- client credentials -> autohosts`. Looks like I got confused about the autohosts table purpose. Currently it's rather empty, but maybe there will be use cases in the future.

> For point 3: I wasn't aware you could change the scopes when refreshing a token. The scope param is ignored when refreshing, you get the same scope as the original token, regardless of the request.

So from https://datatracker.ietf.org/doc/html/rfc6749#section-6

```
scope
         OPTIONAL.  The scope of the access request as described by
         Section 3.3.  The requested scope MUST NOT include any scope
         not originally granted by the resource owner, and if omitted is
         treated as equal to the scope originally granted by the
         resource owner.
```

and later

```
   The authorization server MAY issue a new refresh token (...) If a
   new refresh token is issued, the refresh token scope MUST be
   identical to that of the refresh token included by the client in the
   request.
```

so effectively, you could theoretically reduce the scope of *access* token returned by endpoint, by specifying a subset of specified scopes.

Quite weird, but ok.

geekingfrog commented:
Alright, I think I've got everything now, the last 6 commits should do the job.

I've added some basic check if you provide a `scope` parameter when refreshing a token, but it's incomplete. Since the only scope supported right now is `tachyon.lobby` it doesn't really make sense to support more right now.

p2004a reviewed (approved):
Awesome work! Thank you!

@L-e-x-o-n Please PTAL and merge if all looks good from elixir side.

L-e-x-o-n reviewed (approved):
This is all really well done. 
I just have a few questions, but I think it's ready to be merged .

L-e-x-o-n commented on lib/teiserver/o_auth.ex:
```diff
@@ -0,0 +1,419 @@
+    create_code(user.id, attrs, opts)
+  end
+
+  def create_code(user, attrs, opts) when is_map(user) do
```
Looks like I didn't submit my comment here earlier.
Is this case necessary, shouldn't
```ex
 def create_code(%User{} = user, attrs, opts) do
```
be enough for all cases or am I missing something?

geekingfrog commented on config/runtime.exs:
```diff
@@ -24,13 +24,13 @@ if Teiserver.ConfigHelpers.get_env("PHX_SERVER", nil) do
 end
 
+# used for mailing, checking origins, finding tls certs…
+domain_name = Teiserver.ConfigHelpers.get_env("TEI_DOMAIN_NAME", "beyondallreason.info")
```
Having a default means we can still run the dev server locally without any special config, which is nice.

L-e-x-o-n replied:
Ah, good point.

geekingfrog commented on lib/teiserver/o_auth.ex:
```diff
@@ -0,0 +1,419 @@
+    create_token(user.id, application, opts)
+  end
+
+  def create_token(user, application, opts) when is_map(user) do
+    create_token(user.id, application, opts)
+  end
```
It's used for testing mostly, because the [new_user/2](https://github.com/geekingfrog/teiserver/blob/master/test/support/teiserver_test_lib.ex#L46) function to create test users doesn't return an ecto schema.

And I believe in general, there aren't many case where teiserver directly handle `User` structs. This makes it a bit more robust, especially since we only care about the id.

As a side note, this is something that I'm not a big fan of: the way users are augmented with all sort of data, makes it quite hard to know what's persisted in the DB (and how) from what's transient.

geekingfrog commented on lib/teiserver/o_auth.ex:
```diff
@@ -0,0 +1,419 @@
+      owner_id: user_id,
+      application_id: attrs.id,
+      scopes: attrs.scopes,
+      expires_at: Timex.add(now, Timex.Duration.from_minutes(5)),
```
@p2004a do you have any opinion on this? I'm personally neutral, it's easy enough to change in the code, but does require a recompilation.

p2004a replied:
In my opinion there is no need. The code is supposed to be used immediately by a program, not in any way by humans, so 5m is very generous and I don't think we would need to change it in a hurry. Shortening it also doesn't improve anything from security side.

geekingfrog commented on lib/teiserver/o_auth.ex:
```diff
@@ -0,0 +1,419 @@
+    |> Repo.insert()
+  end
+
+  # TODO: get_valid_token is basically the same as get_valid_code, refactor later
```
I've amended the commit 3f44320e372aa79268a127968c04bbae24f6d4fd
I haven't merged the two functions since they are fairly small and that would be more confusing than anything. The main logic is already extracted into `check_expiry` and the query so I'm happy with how that looks now.

geekingfrog commented on lib/teiserver/o_auth.ex:
```diff
@@ -0,0 +1,419 @@
+      :ok
+    else
+      err -> err
+    end
```
You're right, it's not needed. Commit amended.

geekingfrog commented on lib/teiserver/o_auth.ex:
```diff
@@ -0,0 +1,419 @@
+      compare_hash(challenge, verifier)
+    end
+
+    # valid_verifier(verifier) and :crypto.hash_equals(challenge, verifier)
```
oops, good catch, commit amended

p2004a commented on lib/teiserver/o_auth.ex:
```diff
@@ -0,0 +1,419 @@
+        value: Base.hex_encode32(:crypto.strong_rand_bytes(32), padding: false),
+        application_id: application.id,
+        scopes: application.scopes,
+        expires_at: Timex.add(now, Timex.Duration.from_minutes(30)),
```
This one could be, also the refresh token, but IMHO we don't need to add it already in this PR. I think it is unlikely to change too and can't think about a good reason to have it different across instances.. maaaybe in dev setup for testing clients code.

geekingfrog replied:
I'd say ship it. If we need to tweak that for dev code, that's easy, and this PR is already massive enough.
Worst case, ping me and I can setup a teiserver on my vps.
